### PR TITLE
feat(devexp): Parse and run storage queries

### PR DESCRIPTION
**Pull request by volokluev**
This PR implements querying storages directly at the snuba level. After it is merged, it can be used immediately to query any storage

### Change Summary
* Add a new class `LogicalDataSource` which can be either a Storage or an Entity
    * I wanted to make the parser product clickhouse queries directly but there is too much stuff unrelated to entities that happens in the parser so I kept them separate
    * Eventually I'd actually like to remove `ClickhouseQuery` from the codebase but not in this PR
* Change the snql parser and grammar to be able to parse storage queries directly
* Change the parsing portion of the code to be agnostic to Entity/Storage

### Outstanding tasks

- [x] write more tests for the entity processing stage functionality
- [x] Remove the useless EntityColumnSet class

### CODE REVIEW

File: snuba/query/snql/parser.py [hunk 1]
```diff
@@ -1190,16 +1220,18 @@ def _array_column_conditions(
...
+        _unpack_array_conditions(cast(LogicalQuery, query), from_clause.schema)
     elif isinstance(from_clause, JoinClause):
         alias_map = from_clause.get_alias_node_map()
         for alias, node in alias_map.items():
-            assert isinstance(node.data_source, QueryEntity)  # mypy
-            _unpack_array_conditions(query, node.data_source.schema, alias)
+            if isinstance(node.data_source, QueryEntity):  # mypy
+                _unpack_array_conditions(
+                    cast(LogicalQuery, query), node.data_source.schema, alias
+                )
```
Comment (volokluev) [inline snuba/query/snql/parser.py:1230]: Note to self: this looks kinda weird, make sure it's not something that got lost in the shuffle

File: snuba/query/logical.py [hunk 2]
```diff
@@ -1,15 +1,14 @@
...
-from snuba.query.data_source.simple import Entity
+from snuba.query import LimitBy, OrderBy, ProcessableQuery, SelectedExpression
+from snuba.query.data_source.simple import Entity, LogicalDataSource, Storage
 from snuba.query.expressions import Expression, ExpressionVisitor
 
 
-class Query(AbstractQuery[Entity]):
+class Query(ProcessableQuery[LogicalDataSource]):
     """
     Represents the logical query during query processing.
```
Comment (enochtangg) [inline snuba/query/logical.py:13]: Could we add as a comment that a logical query can either be a logical entity query or a logical storage query for clarity?

**Review by enochtangg — COMMENTED** [review]
Looks great, just a few questions.

File: tests/query/snql/test_storage_query.py [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+            Literal(None, datetime.datetime(2021, 1, 2, 0, 0)),
+        ),
+    ),
+)
+
+
+test_cases = [
+    pytest.param(
+        f"MATCH STORAGE(metric_summaries) SELECT 4-5, trace_id WHERE {added_condition} GRANULARITY 60",
+        StorageQuery(
```
[the remark this answers is not in the document]
Comment (volokluev) [reply tests/query/snql/test_storage_query.py]: It doesn't :). It has been fixed

File: snuba/query/logical.py [hunk 4]
```diff
@@ -81,3 +80,80 @@ def _transform_expressions_impl(
...
+>>> isinstnace(query, EntityQuery):
+>>>     do_something()
+
+These classes give us the best of both worlds by allowing the strict typing and the runtime checking which makes sure
+that when someone is passing in an EntityQuery its datasource is actually an Entity
+
+"""
+
+
+class EntityQuery(Query, ProcessableQuery[Entity], metaclass=_FlexibleQueryType):
```
[the remark this answers is not in the document]
Comment (volokluev) [reply snuba/query/logical.py]: They don't, updated

File: snuba/query/logical.py (hunk 2 above)
Comment (volokluev) [reply snuba/query/logical.py:13]: fosho

**Review by enochtangg — APPROVED** [review]
LGTM after typing and failing tests are fixed

**Pull request merged**